Source/JavaScriptCore/runtime/JSPromise.cpp

```diff
@@ -78,70 +87,81 @@ void JSPromise::visitChildrenImpl(JSCell* cell, Visitor& visitor)
-        return jsUndefined();
-    return internalField(Field::ReactionsOrResult).get();
-}
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
 
-uint32_t JSPromise::flags() const
-{
-    JSValue value = internalField(Field::Flags).get();
-    return value.asUInt32AsAnyInt();
+    auto [promise, resolve, reject] = newPromiseCapability(globalObject, constructor);
+    RETURN_IF_EXCEPTION(scope, { });
+    return createPromiseCapability(vm, globalObject, promise, resolve, reject);
 }
 
-bool JSPromise::isHandled(VM&) const
+JSValue JSPromise::createPromiseCapability(VM& vm, JSGlobalObject* globalObject, JSObject* promise, JSObject* resolve, JSObject* reject)
 {
-    return flags() & isHandledFlag;
+    auto* capability = constructEmptyObject(vm, globalObject->promiseCapabilityObjectStructure());
+    capability->putDirectOffset(vm, promiseCapabilityResolvePropertyOffset, resolve);
+    capability->putDirectOffset(vm, promiseCapabilityRejectPropertyOffset, reject);
+    capability->putDirectOffset(vm, promiseCapabilityPromisePropertyOffset, promise);
+    return capability;
 }
 
-JSValue JSPromise::createNewPromiseCapability(JSGlobalObject* globalObject, JSPromiseConstructor* promiseConstructor)
+std::tuple<JSObject*, JSObject*, JSObject*> JSPromise::newPromiseCapability(JSGlobalObject* globalObject, JSValue constructor)
 {
     VM& vm = globalObject->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    JSFunction* newPromiseCapabilityFunction = globalObject->newPromiseCapabilityFunction();
-    auto callData = JSC::getCallData(newPromiseCapabilityFunction);
-    ASSERT(callData.type != CallData::Type::None);
+    if (constructor == globalObject->promiseConstructor()) {
+        auto* promise = JSPromise::create(vm, globalObject->promiseStructure());
+        auto [resolve, reject] = promise->createFirstResolvingFunctions(vm, globalObject);
+        return { promise, resolve, reject };
+    }
 
-    MarkedArgumentBuffer arguments;
-    arguments.append(promiseConstructor);
-    ASSERT(!arguments.hasOverflowed());
-    RELEASE_AND_RETURN(scope, call(globalObject, newPromiseCapabilityFunction, callData, jsUndefined(), arguments));
-}
+    if (constructor == globalObject->internalPromiseConstructor()) {
+        auto* promise = JSInternalPromise::create(vm, globalObject->internalPromiseStructure());
+        auto [resolve, reject] = promise->createFirstResolvingFunctions(vm, globalObject);
+        return { promise, resolve, reject };
+    }
 
-JSPromise::DeferredData JSPromise::convertCapabilityToDeferredData(JSGlobalObject* globalObject, JSValue promiseCapability)
-{
-    DeferredData result;
-    VM& vm = globalObject->vm();
-    auto scope = DECLARE_THROW_SCOPE(vm);
+    auto* executor = JSFunctionWithFields::create(vm, globalObject, vm.promiseCapabilityExecutorExecutable(), 2, nullString());
+    executor->fields()[0].set(vm, executor, jsUndefined());
+    executor->fields()[1].set(vm, executor, jsUndefined());
```

Comment:
We should really have these be named in some way. It's consistently been a long term matainance burden when these are just offsets and not named fields. Even just having an enum we consistently use would be fine.

Reply:
Done